src/script/util/E2EIdentity/DelayTimer.ts

```diff
@@ -0,0 +1,279 @@
+/*
+ * Wire
+ * Copyright (C) 2023 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import {TaskScheduler} from '@wireapp/core/lib/util/TaskScheduler';
+
+import {getLogger} from 'Util/Logger';
+
+import {getDelayTime} from './helper/delay';
+import {DelayTimerStore} from './helper/storage';
+
+const logger = getLogger('E2EIdentity.DelayTimer');
+
+interface CreateGracePeriodTimerParams {
+  gracePeriodInMS: number;
+  gpCallback: () => void;
+  delayCallback: () => void;
+}
+
+class GracePeriodTimer {
+  private static instance: GracePeriodTimer | null = null;
+  private gracePeriodInMS: number;
+  private gpCallback: () => void;
+  private delayCallback: () => void;
+  private delayPeriodTimerKey: string = 'E2EIdentity_DelayTimer';
+  private gracePeriodTimerKey: string = 'E2EIdentity_GracePeriodTimer';
+
+  private constructor({gracePeriodInMS, gpCallback, delayCallback}: CreateGracePeriodTimerParams) {
+    this.gracePeriodInMS = gracePeriodInMS;
+    this.gpCallback = gpCallback;
+    this.delayCallback = delayCallback;
+    this.initialize();
+  }
+
+  /**
+   * Get the singleton instance of GracePeriodTimer or create a new one
+   * For the first time, params are required to create the instance
+   * After that, params are optional and can be used to update the grace period timer
+   * @param params The params to create the grace period timer
+   * @returns The singleton instance of GracePeriodTimer
+   */
+  public static getInstance(params?: CreateGracePeriodTimerParams) {
+    if (!GracePeriodTimer.instance) {
+      if (!params) {
+        throw new Error('GracePeriodTimer is not initialized. Please call getInstance with params.');
+      }
+      GracePeriodTimer.instance = new GracePeriodTimer(params);
+    }
+    return GracePeriodTimer.instance;
+  }
+
+  /**
+   * @param CreateGracePeriodTimerParams The params to create the grace period timer
+   */
+  public updateParams({gracePeriodInMS, gpCallback, delayCallback}: CreateGracePeriodTimerParams) {
+    DelayTimerStore.clear.all();
+    this.clearGracePeriodTimer();
+    this.clearDelayPeriodTimer();
+    this.gracePeriodInMS = gracePeriodInMS;
+    this.gpCallback = gpCallback;
```

Comment:
gpCallback doesn't clearly tell us the purpose of it, can you give a more descriptive name?

Reply:
Indeed, I should follow my own preferences here and call it sth like gracePeriodExpiredCallback